File: src/auth/password.ts
import { randomBytes, scrypt, timingSafeEqual } from 'node:crypto'

const SCRYPT_N = 16384
const SCRYPT_R = 8
const SCRYPT_P = 1
const KEY_LEN = 64
const SALT_LEN = 32

/**
 * Hash a password using Node's built-in scrypt.
 * Returns a self-describing string: `scrypt:N=...,r=...,p=...:salt_hex:hash_hex`
 */
export async function hashPassword(plain: string): Promise<string> {
  const salt = randomBytes(SALT_LEN)
  const derived = await new Promise<Buffer>((resolve, reject) => {
    scrypt(plain, salt, KEY_LEN, { N: SCRYPT_N, r: SCRYPT_R, p: SCRYPT_P }, (err, key) => {
      if (err) reject(err)
      else resolve(key)
    })
  })
  return `scrypt:N=${SCRYPT_N},r=${SCRYPT_R},p=${SCRYPT_P}:${salt.toString('hex')}:${derived.toString('hex')}`
}

/**
 * Verify a password against a hash string produced by `hashPassword`.
 * Uses timing-safe comparison.
 */
export async function verifyPassword(plain: string, stored: string): Promise<boolean> {
  const parts = stored.split(':')
  if (parts[0] !== 'scrypt' || parts.length !== 4) return false

  const salt = Buffer.from(parts[2], 'hex')
  const expected = Buffer.from(parts[3], 'hex')

  const derived = await new Promise<Buffer>((resolve, reject) => {
    scrypt(plain, salt, expected.length, { N: SCRYPT_N, r: SCRYPT_R, p: SCRYPT_P }, (err, key) => {
      if (err) reject(err)
      else resolve(key)
    })
  })

  return timingSafeEqual(derived, expected)
}

File: internal/auth/password.go
// Package auth verifies login credentials.
package auth

import (
	"crypto/subtle"
	"encoding/hex"
	"fmt"
	"strings"

	"golang.org/x/crypto/scrypt"
)

// Stored hashes carry N/r/p in their prefix, but the contract is to verify
// with these fixed constants regardless of what the string says.
const (
	scryptN      = 16384
	scryptR      = 8
	scryptP      = 1
	scryptKeyLen = 64
)

func VerifyPlaintext(password, configured string) bool {
	return subtle.ConstantTimeCompare([]byte(password), []byte(configured)) == 1
}

// VerifyScrypt checks a password against the self-describing stored form
// "scrypt:N=16384,r=8,p=1:salt_hex:hash_hex".
func VerifyScrypt(password, stored string) (bool, error) {
	parts := strings.Split(stored, ":")
	if len(parts) != 4 || parts[0] != "scrypt" {
		return false, fmt.Errorf("unrecognized password hash format")
	}
	salt, err := hex.DecodeString(parts[2])
	if err != nil {
		return false, fmt.Errorf("decoding salt: %w", err)
	}
	want, err := hex.DecodeString(parts[3])
	if err != nil {
		return false, fmt.Errorf("decoding hash: %w", err)
	}
	derived, err := scrypt.Key([]byte(password), salt, scryptN, scryptR, scryptP, scryptKeyLen)
	if err != nil {
		return false, err
	}
	return subtle.ConstantTimeCompare(derived, want) == 1, nil
}
